### PR TITLE
Error parsing DateTime and Boolean fields on mutation

### PR DESCRIPTION
**Pull request by klaasvandeweerdt**

### CODE REVIEW

**Comment by klaasvandeweerdt** [comment]
@bennettgo , @den-sergeyev-workato 
I pushed another update. There was also an issue, in the same lambda, to include a boolean `false` value. A `false` value was excluded from the query, while `true` was included. I also updated the specs to include this specific scenario.

**Comment by klaasvandeweerdt** [comment]
@den-sergeyev-workato 
I added another fix for string parsing. It failed when using specific characters that require escaping in JSON.

This is an input example.
```
{
	"account": "wdc",
	"object": "requestUpdate{}",
	"argument_input": {
		"id": "Nl1...OTgy",
		"note": "**1\\. \" Bold**"
	}
}
```
I updated the code at line [2316](https://github.com/Workato-Partners/4me/blob/ba429b9bec381cff20ca4809a73510c0a172c054/connector.rb#L2316), changed it from `value = "\"#{value}\""` to `value = value.to_json` which solves the issue. I explicitly mention this as the same issue may exist in the generic GraphQL connector.

File: connector.rb [hunk 1]
```diff
@@ -2307,12 +2307,13 @@
       end
 
       # treat blank string for optional fields as nil
-      value = field['default'] if value.blank? && !optional
+      value = nil if value.is_a?(String) && optional && value == ''
```
Comment (den-sergeyev-workato) [inline connector.rb]: Can you explain why it is not enough to have these 2 conditions?
```
value = nil if optional && value == ''
```

File: spec/actions/mutation_spec.rb [hunk 2]
```diff
@@ -28,4 +28,19 @@
...
+    let(:input) { JSON.parse(File.read('fixtures/actions/mutation/input/request_create_with_error.json')) }
+    subject(:output) { connector.actions.mutation(input) }
+
+    it 'contains a response' do
+      expect(output).to be_present
+    end
+
+    it 'contains errors' do
+      expect(output).to include('errors')
+      puts output
```
Comment (den-sergeyev-workato) [inline spec/actions/mutation_spec.rb]: Do we need puts here?

File: connector.rb [hunk 3]
```diff
@@ -2307,12 +2307,13 @@
...
 
-      if value.present?
-        if %w[ISO8601Timestamp ISO8601DateTime ISO8601Date].include?(type) ||
-           (type == 'string' && control_type != 'select')
-          value = "\"#{value}\""
+      unless value.nil? && optional
+        value = field['default'] if value.nil?
+
+        if %w[date_time date].include?(type) || (type == 'string' && control_type != 'select')
```
Comment (den-sergeyev-workato) [inline connector.rb:2315]: What changed so we change `ISO8601Timestamp, ISO8601DateTime, ISO8601Date` to `date_time date`?

Comment (den-sergeyev-workato) [inline connector.rb:2315]: What are possible options for control_type with type string?

**Review by den-sergeyev-workato — COMMENTED** [review]
Overall LGTM, some questions to clarify though.

I'm wondering why do we have so many changes in tapes, is it all affected by change?
from 
```"\"#{value}\""``` 
to 
```value.to_json```

Can you, please, link or description to first bug what this PR fixes?

File: spec/actions/mutation_spec.rb (hunk 2 above)
Comment (klaasvandeweerdt) [reply spec/actions/mutation_spec.rb]: Sorry, still left from troubleshooting. I removed it.

File: connector.rb (hunk 3 above)
Comment (klaasvandeweerdt) [reply connector.rb:2315]: A bug. `date` and `date_time` were not serialized properly (not encapsulated) as the comparison was wrong. `ISO8601Timestamp`, `ISO8601DateTime`, `ISO8601Date` are GraphQL types.

File: connector.rb (hunk 1 above)
Comment (klaasvandeweerdt) [reply connector.rb]: Optional string values should not be part of the GraphQL payload if they are empty, they get defaulted to `nil`. They are excluded in the next condition `unless value.nil? && optional`.

Comment (den-sergeyev-workato) [reply connector.rb]: Got it, my question was are these two checks are equivalent? They are always false if value is not string equal to ""

```
value = nil if value.is_a?(String) && optional && value == ''
```
vs
```
value = nil if optional && value == ''
```

Comment (klaasvandeweerdt) [reply connector.rb]: I copied it over from the universal GraphQL connector.
Updated it as you suggested as I cannot imagine a scenario where equal operator overload would return `''` for a non-string value.

File: connector.rb (hunk 3 above)
Comment (klaasvandeweerdt) [reply connector.rb:2315]: If the type is a `select` then the value is an enum and should not be serialized.

Comment (den-sergeyev-workato) [reply connector.rb:2315]: Got it, I mean is it the only control_type we should not serialize? Just want to double check.

**Comment by klaasvandeweerdt** [comment]
@den-sergeyev-workato, there are 3 bug fixes.
1. The initial PR was for the incorrect parsing of `date_time` and `date` values, not encapsulated.
2. The `format_arg` lambda did not serialize a `false` value
3. The `format_arg` lambda did not properly escape special characters, changed it to `value.to_json` instead of `"\"#{value}\""`

I deleted the existing tape libraries, as all action tests were failing after the `value.to_json`. But also because our GraphQL API gets updated almost weekly and I wanted to make sure the recordings were updated.

**Review by den-sergeyev-workato — APPROVED** [review]

**Pull request merged**